### PR TITLE
hex-literal: remove support for comments inside literals and migrate to CTFE

### PR DESCRIPTION
**Pull request by newpavlov**
Potential breaking change for v1.0, see  #814.

### CODE REVIEW

**Comment by newpavlov** [comment]
As [suggested](https://github.com/RustCrypto/utils/issues/814#issuecomment-1438272397) by @Nugine, I've migrated this crate from proc macro to CTFE. Initially I thought it would require Nightly, but looking closer at the `const_str` crate, it can work on stable as well. It result in a lighter crate, which likely should compile significantly faster.

**Comment by newpavlov** [comment]
@tarcieri 
I think we can release these changes as v0.4. WDYT?

**Comment by tarcieri** [comment]
@newpavlov sure!

**Review by tarcieri — APPROVED** [review]

**Comment by triamazikamno** [comment]
Hi, this change seem to have broken the usage of hex! in the match patterns, i.e.
```
match buf {
hex!("0a0b") => {}
```
throws now:
```
error: expected pattern, found `{`
   |
60 |         hex!("0a0b") => {
   |         ^^^^^^^^^^^^^^^^
   |         |
   |         expected pattern
   |         in this macro invocation
   |         this macro call doesn't expand to a pattern
```

**Comment by newpavlov** [comment]
Unfortunately, I think it's inevitable with the CTFE approach until [implementation](https://github.com/rust-lang/rust/issues/76001) of RFC 2920 will be stabilized. You can either continue to use v0.3, or introduce named constants and use them in patterns.

**Comment by triamazikamno** [comment]
I see, thank you. Will stick to the v0.3 for now

**Comment by tarcieri** [comment]
Another small regression (which is easy enough to work around, but was convenient before):

https://github.com/RustCrypto/SSH/actions/runs/4599840083/jobs/8125721544?pr=92#step:5:13

```
error[E0716]: temporary value dropped while borrowed
 --> ssh-encoding/tests/decode.rs:8:21
  |
8 |     let mut bytes = hex!("42").as_slice();
  |                     ^^^^^^^^^^           - temporary value is freed at the end of this statement
  |                     |
  |                     creates a temporary value which is freed while still in use
9 |     let ret = u8::decode(&mut bytes).unwrap();
  |                          ---------- borrow later used here
```

**Comment by newpavlov** [comment]
Yes, it's unfortunate. Though in such cases I usually use shorter `let mut bytes = &hex!("...");` with hinted type, if necessary.

File: hex-literal/src/lib.rs [hunk 1]
```diff
@@ -1,169 +1,86 @@
...
-            unexpected => panic!("expected string literal, got `{}`", unexpected),
-        };
-        out_ts.extend(iter);
-    }
-    TokenStream::from(TokenTree::Group(Group::new(Delimiter::Bracket, out_ts)))
+#[macro_export]
+macro_rules! hex {
+    ($($s:literal)*) => {{
+        const STRINGS: &[&'static [u8]] = &[$($s.as_bytes(),)*];
+        $crate::decode::<{ $crate::len(STRINGS) }>(STRINGS)
```
Comment (Nugine) [inline hex-literal/src/lib.rs:84]: This line may not be const-promoted (or const-evaluated), which may causes performance regression.

**Pull request merged**